### ISSUE TITLE
Add support for unions

### ISSUE THREAD

**Issue by sdroege** [gtk-rs/gir#334]
Blocked by https://github.com/rust-lang/rust/issues/32836

Unfortunately needed for handling various types from Gst-1.0.gir

**Comment by EPashkin** [comment]
Can you give more info? gir from http://packages.ubuntu.com/trusty-updates/amd64/libgstreamer1.0-dev
parsed normally in sys mode.

**Comment by sdroege** [comment]
It fails here with:
```
caused by: 9133:7 src/parser.rs:206 Unexpected element <union>
```

This is with GStreamer 1.11.2. Problem is this: https://github.com/GStreamer/gstreamer/blob/master/gst/gstcontrolbinding.h#L77
(Unfortunately there are more cases like this)

Also the version from trusty is ancient, but that's probably unrelated here.

**Comment by sdroege** [comment]
Support for unions will also be needed to properly be able to handle e.g. GMutex in structs (GMutex is a union itself).

**Comment by sdroege** [comment]
And I should add that the problem here is not that parsing fails. The problem is that the generated Rust structs are truncated whenever they contain unions, and as such they're semi-useless :)

**Comment by EPashkin** [comment]
Rechecked it with zesty's 1.10, gir really can't parse these.

**Comment by sdroege** [comment]
Without direct support for unions in Rust, what could probably be done in many cases (probably all GStreamer cases at least) is to check the contents of the union and then just put some dummy item in place of the union that has the same size as the union.

Also for GMutex you could make it a "struct GMutex(u64)", which should cover all platforms currently supported by GLib. (Fortunately in GStreamer the cases are even simpler)

**Comment by Luke-Nukem** [comment]
What is the current status on this? I'm in need of `GIArgument` which is a union. At the moment I'm implementing it as an untagged union in nightly rust for experimenting. Not sure how it will behave as of yet - but it is on track to becoming stabilised by either rustc-1.19 or 1.20.

`GIArgument` is currently generated as a `c_void`, which means there are no members I can access through Rust.

**Comment by sdroege** [comment]
Nobody is working on it AFAIK, and the consensus IIRC was to wait for the Rust union feature instead of implementing something homegrown here. And for the types where it is really needed now already (GMutex, GValue, etc) just implement it manually for now.

That said, union support is in nightly so maybe this could be implemented already behind some configuration flag.

**Comment by tmiasko** [comment]
Since #415 the feature-gating of unions is removed, so I suppose we can close this now?
The remaining problems with unions are essentially those related to bitfields which are tracked in separate issue #465.
